quickwit/quickwit-indexing/src/source/gcp_pubsub_source.rs

```diff
@@ -0,0 +1,241 @@
+use serde_json::Value as JsonValue;
+use tokio::time;
+use tracing::{info, warn};
+
+use super::SourceActor;
+use crate::actors::DocProcessor;
+use crate::source::{
+    BatchBuilder, Source, SourceContext, SourceExecutionContext, TypedSourceFactory,
+};
+
+const BATCH_NUM_BYTES_LIMIT: u64 = 5_000_000;
+const DEFAULT_MAX_MESSAGES_PER_PULL: i32 = 1000;
+// const DEFAULT_PULL_PARALLELISM: u64 = 10; // TODO: is 10 too high as a default?
+
+pub struct GcpPubSubSourceFactory;
+
+#[async_trait]
+impl TypedSourceFactory for GcpPubSubSourceFactory {
+    type Source = GcpPubSubSource;
+    type Params = GcpPubSubSourceParams;
+
+    async fn typed_create_source(
+        ctx: Arc<SourceExecutionContext>,
+        params: GcpPubSubSourceParams,
+        _checkpoint: SourceCheckpoint, // TODO: Use checkpoint!
+    ) -> anyhow::Result<Self::Source> {
+        GcpPubSubSource::try_new(ctx, params).await
+    }
+}
+
+#[derive(Default)]
+pub struct GcpPubSubSourceState {
+    subscription_is_empty: bool,
+    doc_count: u64,
+}
+
+pub struct GcpPubSubSource {
+    ctx: Arc<SourceExecutionContext>,
+    subscription: String,
+    state: GcpPubSubSourceState,
+    subscription_source: Subscription,
+    backfill_mode_enabled: bool,
+    partition_id: String,
+    // Parallelism will be enabled in next PR
+    // pull_parallelism: u64,
+    max_messages_per_pull: i32,
+}
+
+impl GcpPubSubSource {
+    pub async fn try_new(
+        ctx: Arc<SourceExecutionContext>,
+        params: GcpPubSubSourceParams,
+    ) -> anyhow::Result<Self> {
+        let subscription = params.subscription.clone();
+        let backfill_mode_enabled = params.enable_backfill_mode;
+        // let pull_parallelism = params.pull_parallelism.unwrap_or(DEFAULT_PULL_PARALLELISM);
+        let max_messages_per_pull = match params.max_messages_per_pull {
+            Some(max) => max,
+            None => DEFAULT_MAX_MESSAGES_PER_PULL,
+        };
+
+        let config = match params.credentials {
+            Some(_credentials) => todo!("Add specific credentials file config"),
+            None => ClientConfig::default().with_auth(),
+        }
+        .await
+        .expect("Failed to use GCP credentials");
+
+        let client = Client::new(config)
+            .await
+            .expect("Failed to create GcpPubSub client");
+        let subscription_source = client.subscription(&subscription);
+
+        info!(
+            index_id=%ctx.index_uid.index_id(),
+            source_id=%ctx.source_config.source_id,
+            subscription=%subscription,
+            // parallelism=%pull_parallelism,
+            "Starting GcpPubSub source."
+        );
+
+        Ok(GcpPubSubSource {
+            ctx,
+            subscription,
+            state: GcpPubSubSourceState::default(),
+            subscription_source,
+            backfill_mode_enabled,
+            // TODO: get the real value
+            partition_id: "<node_id>/<pipeline_ord>".to_string(),
```

Comment:
@guilload can you give some guidance where I can get those information ? like you suggested in https://github.com/quickwit-oss/quickwit/pull/3605#issuecomment-1668563149

Reply:
Yes, a refactor of `SourceExecutionContext` is required to have access to those parameters. Let me push a PR before the end of the week.

Reply:
I updated my code, I put `append_random_suffix` to not be block by it. I can add those parameter after the refactor is done

Reply:
Sounds good.